### Contents
##### 1.0 Importing of libraries
##### 2.0 Importing of orders dataset
##### 3.0 Checking the information of the dataframe
##### 4.0 Changing of data type
##### 5.0 Changing of unintuitive column names
##### 6.0 Calculating the Busiest hour 
##### 7.0 Importing of the department dataset
##### 8.0 Tranposing of the department dataframe and creating new header 
##### 9.0  Creating dictionary of the new wrangled department dataframe
##### 10 Importing Product dataset
##### 11 Creating subset using isin


### 1.0 Importing of libraries

In [3]:
#importing of libraries
import pandas as pd
import numpy as np
import os

### 2.0 Importing of orders dataset

In [118]:
df_ords = pd.read_csv('Downloads/CF_orders.csv', index_col = False)

In [119]:
# dataframe heads
df_ords.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

### 3.0 Checking the information of the dataframe

In [120]:
# Order information
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                int64  
 1   user_id                 int64  
 2   eval_set                object 
 3   order_number            int64  
 4   order_dow               int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(5), object(1)
memory usage: 182.7+ MB


##### QUESTION 2 :Find another identifier variable in the df_ords dataframe that doesn’t need to be included in your analysis as a numeric variable and change it to a suitable format.




### 4.0 Changing of data type

In [121]:
# changing the order_id columns data type to string 
df_ords1 = df_ords['order_id'].astype('str')
df_ords1.head()

0    2539329
1    2398795
2     473747
3    2254736
4     431534
Name: order_id, dtype: object

 ###### The order_id data type is an integer and changed to string 

### 5.0 Changing of unintuitive column names

######  QUESTION 3 : Look for a variable in your df_ords dataframe with an unintuitive name and change it name without overwriting the data frame

In [122]:
df_ords_col_change = df_ords.rename(columns = {'order_dow': 'order_of_the _week'})
df_ords_col_change.head()

order_id  user_id eval_set  order_number  order_of_the _week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

### 6.0 Calculating the Busiest hour 

###### Question 4:Your client wants to know what the busiest hour is for placing orders. Find the frequency of the corresponding variable and share your findings.

In [123]:
# using value_count to find the busiest hour
df_ords['order_hour_of_day'].value_counts().head()

10    288418
11    284728
15    283639
14    283042
13    277999
Name: order_hour_of_day, dtype: int64

 The busiest hour is 10:am with the highest count of 288418

### 7.0 Importing of the department dataset

In [6]:
df_dep = pd.read_csv('/Users/shaqurcasco/Downloads/ Instacart Basket June: 26: 2023 CF/3 SCRIPT/CF Achievement 4/CF_departments.csv', index_col = False)
df_dep

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

### 8.0 Tranposing of the department dataframe and creating new header 

In [7]:
# transposing the department dataframe
df_dep.T.head()

0
department_id  department
1                  frozen
2                   other
3                  bakery
4                 produce

In [8]:
# Resetting the index of the transpose dataframe
df_dep.T.reset_index().head()

index           0
0  department_id  department
1              1      frozen
2              2       other
3              3      bakery
4              4     produce

In [9]:
# Getting the header of the transpose dataframe 
new_header = df_dep.T.iloc[0]
new_header

0    department
Name: department_id, dtype: object

In [10]:
#  Indexing the dataframne of the department starting from 1:
df_dep.T_new = df_dep.T.iloc[1:]

df_dep.T_new.head()

/var/folders/9f/n36r4f3s7qsf6wgq534gqhbh0000gn/T/ipykernel_6251/3363352293.py:1: UserWarning: Pandas doesn't allow columns to be created via a new attribute name - see https://pandas.pydata.org/pandas-docs/stable/indexing.html#attribute-access
  df_dep.T_new = df_dep.T.iloc[1:]


0
1   frozen
2    other
3   bakery
4  produce
5  alcohol

In [11]:
# Assigning the header created to the new dataframe created
df_dep.T_new.columns = new_header
df_dep.T_new.head()

department_id department
1                 frozen
2                  other
3                 bakery
4                produce
5                alcohol

In [13]:
# Saving the new transpose department dataframe as CSV
df_dep.T_new.to_csv('/Users/shaqurcasco/Downloads/ Instacart Basket June: 26: 2023 CF/3 SCRIPT/CF Achievement 4/department_wrangling.csv')

### 9.0  Creating dictionary of the new wrangled department dataframe                             

In [65]:
# create dictionary of wrangled department dataframe
dic = df_dep.T_new.to_dict('index')

In [66]:
dic

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

#### QUESTION 5. Determine the meaning behind a value of 4 in the "department_id" column within the df_prods dataframe using a data dictionary

In [68]:
# Getting the department of the value 4
print(dic.get('4'))

{'department': 'produce'}


### 10 Importing Product dataset 

In [70]:
# importing the product dataset
df_prod = pd.read_csv('Downloads/CF_products.csv')
df_prod.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [72]:
# Getting the produce dataframe with department id = 4
df_produce = df_prod.loc[df_prod["department_id"]==4]
df_produce.head()

product_id                                      product_name  aisle_id  \
30             31                                White Pearl Onions       123   
42             43                               Organic Clementines       123   
44             45                                 European Cucumber        83   
65             66                         European Style Spring Mix       123   
88             89                    Yogurt Fruit Dip Sliced Apples       123   
...           ...                                               ...       ...   
49582       49578                                Black Garlic Bulbs       123   
49623       49619                                        Opo Squash        83   
49639       49635  Baby Food Blueberry, Parsnip & Buckwheat Stage 2        83   
49661       49657                                 Cabernet Tomatoes        83   
49687       49683                                    Cucumber Kirby        83   

       department_id  prices  
30                 4     7.5  
42                 4    11.5  
44                 4    14.3  
65                 4    11.7  
88                 4    12.6  
...              ...     ...  
49582              4     8.0  
49623              4    12.7  
49639              4    12.5  
49661              4     8.3  
49687              4    13.2  

[1684 rows x 5 columns]

### 11 Creating subset using isin

Question 6: The sales team in your client’s organization wants to know more about breakfast item sales. 
Create a subset containing only the required information.

In [74]:
Creating subset using isin() with department id = 4
df_breakfast = df_prod.loc[df_prod["department_id"].isin([4])]
df_breakfast.head()

product_id                    product_name  aisle_id  department_id  \
30          31              White Pearl Onions       123              4   
42          43             Organic Clementines       123              4   
44          45               European Cucumber        83              4   
65          66       European Style Spring Mix       123              4   
88          89  Yogurt Fruit Dip Sliced Apples       123              4   

    prices  
30     7.5  
42    11.5  
44    14.3  
65    11.7  
88    12.6

###### QUESTION 7 : They’d also like to see details about products that customers might use to throw dinner parties.Your task is to find all observations from the entire dataframe that include items from the following departments: alcohol, deli, beverages, and meat/seafood. You’ll need to present this subset to your client.

In [76]:
# Listing subset for the dinner party with department id = 5,7,12,20
df_dinnerparties = df_prod.loc[df_prod['department_id'].isin([5,7,12,20])]
df_dinnerparties.head()

product_id                                    product_name  aisle_id  \
2            3            Robust Golden Unsweetened Oolong Tea        94   
6            7                  Pure Coconut Water With Orange        98   
9           10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10          11                               Peach Mango Juice        31   
16          17                               Rendered Duck Fat        35   

    department_id  prices  
2               7     4.5  
6               7     4.4  
9               7     8.4  
10              7     2.8  
16             12    17.1

##### Question 8 It’s important that you keep track of total counts in your dataframes. How many rows does the last dataframe you created have?

In [77]:
#checking the columns a d rows of the subset dinner-party 
df_dinnerparties.shape

(7650, 5)

 The number of row is 7650 and row is 5

##### Questuion 9 Someone from the data engineers team in Instacart thinks they’ve spotted something strange about the customer with a "user_id" of “1.”Extract all the information you can about this user.

In [80]:
# Listing subset for the dinner party with department id = 1 and assigning it as df_customer 
df_customer = df_ords.loc[df_ords['user_id'].isin([1])]
df_customer.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

##### Question 10: You also need to provide some details about this user’s behavior. What basic stats can you provide based on the information you have?

In [81]:
# checking the statistics of the df_customer 
df_customer.describe()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
count  1.100000e+01     11.0     11.000000  11.000000          11.000000   
mean   1.923450e+06      1.0      6.000000   2.636364          10.090909   
std    1.071950e+06      0.0      3.316625   1.286291           3.477198   
min    4.315340e+05      1.0      1.000000   1.000000           7.000000   
25%    8.690170e+05      1.0      3.500000   1.500000           7.500000   
50%    2.295261e+06      1.0      6.000000   3.000000           8.000000   
75%    2.544846e+06      1.0      8.500000   4.000000          13.000000   
max    3.367565e+06      1.0     11.000000   4.000000          16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

### Task Step 12 - Exporting df_ords

In [ ]:
# Saving the new transpose department dataframe as CSV
df_dep.T_new.to_csv('/Users/shaqurcasco/Downloads/ Instacart Basket June: 26: 2023 CF/3 SCRIPT/CF Achievement 4/orders_wrangled.csv')

### Task Step 12 - Exporting df_dep_T_new

In [ ]:
# Saving the new transpose department dataframe as CSV
df_dep.T_new.to_csv('/Users/shaqurcasco/Downloads/ Instacart Basket June: 26: 2023 CF/3 SCRIPT/CF Achievement 4/department_wrangling.csv')